docs/decisions/0022-planner-telemetry-enhancement.md

```diff
@@ -0,0 +1,225 @@
+---
+status: { proposed }
+contact: { TaoChenOSU }
+date: { 2023-11-18 }
+deciders: alliscode, dmytrostruk, markwallace, SergeyMenshykh, stephentoub
+consulted: {}
+informed: {}
+---
+
+# Planner Telemetry Enhancement
+
+## Context and Problem Statement
+
+It would be extremely beneficial for applications using Semantic Kernel' planning features to be able to continuously monitor the performance of planners and plans.
+
+## Scenarios
+
+Contoso is a company that is developing an AI application using SK.
+
+1. Contoso needs to continuously monitor the token usage of a particular planner, including prompt tokens, completion tokens, and the total tokens.
+2. Contoso needs to continuously monitor the time it takes for a particular planner to create a plan.
+3. Contoso needs to continuously monitor the success rate of a particular planner in creating a valid plan.
+4. Contoso needs to continuously monitor the success rate of a particular plan type being executed successfully.
+5. Contoso wants to be able to see the token usage of a particular planner run.
+6. Contoso wants to be able to see the time taken to create a plan of a particular planner run.
+7. Contoso wants to be able to see the steps in a plan.
+8. Contoso wants to be able to see the inputs&outputs of each plan step.
+9. Contoso wants to change a few settings that may affect the performance of the planners. They would like to know how the performance will be affected before committing the changes.
+10. Contoso wants to update to a new model that is cheaper and faster. They would like to know how the new model performs in planning tasks.
+
+## Out of scope
+
+1. We focus on Application Insights. Although other telemetry service options are supported technically, we will not cover possible ways of setting them up in this ADR.
+2. This ADR does not seek to modify the current instrumentation design in SK.
+
+## Decision Drivers
+
+- The framework should be telemetry service agnostic.
+- Enabling and disabling specific telemetry items should require minimum effort.
+
+## Considered Options
+
+- Function hooks
+  - Inject logic to functions that will get executed before or after a function is invoked.
+- Distributed tracing
+  - Activity tags & Metrics
+    - Use activities to record individual operations.
+    - Use tags in activities to record custom information.
+    - Use baggage in activities to more easily correlate dependencies.
```

Comment:
baggage is probably the wrong thing to use here...

Reply:
Could you elaborate?

Reply:
Discussed offline.